crates/bevy_ecs/src/storage/blob_vec.rs

```diff
@@ -0,0 +1,371 @@
+                    array_layout(&self.item_layout, self.capacity)
+                        .expect("array layout should be valid"),
+                    self.item_layout.size() * new_capacity,
+                )
+            };
+
+            self.data = UnsafeCell::new(NonNull::new(new_data).unwrap_or_else(|| {
+                handle_alloc_error(
+                    array_layout(&self.item_layout, new_capacity)
+                        .expect("array layout should be valid"),
+                )
+            }));
+        }
+        self.capacity = new_capacity;
+    }
+
+    /// # Safety
+    /// `index` must be in bounds
+    /// Allows aliased mutable access to `index`'s data. Caller must ensure this does not happen
+    #[inline]
+    pub unsafe fn set_unchecked(&self, index: usize, value: *mut u8) {
+        let ptr = self.get_unchecked(index);
+        std::ptr::copy_nonoverlapping(value, ptr, self.item_layout.size());
+    }
+
+    /// increases the length by one (and grows the vec if needed) with uninitialized memory and returns the index
+    /// # Safety
+    /// the newly allocated space must be immediately populated with a valid value
+    #[inline]
+    pub unsafe fn push_uninit(&mut self) -> usize {
+        self.reserve(1);
+        let index = self.len;
+        self.len += 1;
+        index
+    }
+
+    /// # Safety
+    /// len must be <= capacity. if length is decreased, "out of bounds" items must be dropped. Newly added items must be
+    /// immediately populated with valid values and length must be increased. For better unwind safety, call [BlobVec::set_len]
+    /// _after_ populating a new value.
+    pub unsafe fn set_len(&mut self, len: usize) {
+        self.len = len;
+    }
+
+    /// Performs a "swap remove" at the given `index`, which removes the item at `index` and moves the last item
+    /// in the [BlobVec] to `index` (if `index` is not the last item). It is the caller's responsibility to
+    /// drop the returned pointer, if that is desirable.
+    /// # Safety
+    /// It is the caller's responsibility to ensure that `index` is < self.len()
+    /// Callers should _only_ access the returned pointer immediately after calling this function.
+    #[inline]
+    pub unsafe fn swap_remove_and_forget_unchecked(&mut self, index: usize) -> *mut u8 {
+        let last = self.len - 1;
+        let swap_scratch = (*self.swap_scratch.get()).as_ptr();
+        std::ptr::copy_nonoverlapping(
+            self.get_unchecked(index),
+            swap_scratch,
+            self.item_layout.size(),
+        );
+        std::ptr::copy(
+            self.get_unchecked(last),
+            self.get_unchecked(index),
+            self.item_layout.size(),
+        );
+        self.len -= 1;
+        swap_scratch
+    }
+
+    /// # Safety
+    /// index must be in-bounds
+    #[inline]
+    pub unsafe fn swap_remove_and_drop_unchecked(&mut self, index: usize) {
+        let value = self.swap_remove_and_forget_unchecked(index);
```

Comment:
Can methods with easily verifiable safety conditions like this method get a `debug_assert!` for the safety condition? In this case `debug_assert!(index < self.len);`.

Reply:
Good call. Funny story: adding a debug_assert for index validity to BlobVec::get_unchecked causes a panic in debug mode when dropping BlobVec. This is because we now set `self.len = 0` _before_ calling `self.get_unchecked(index)` for each item (for unwind safety). I guess the fix is to manually offset the pointer in `clear()` so we can still use the debug_assert elsewhere.